server/server.cpp

```diff
@@ -38,64 +82,38 @@ int main(int argc, char **argv) {
     return 1;
   }
   // 3 listen
-  int listen_value = listen(sock, 1);
+  int listen_value = listen(sock, 10);
   if (listen_value < 0) {
 
     perror("Could not listen");
     return 1;
   }
 
-  // 4 accept
-  struct sockaddr_in remote_address;
-  memset(&remote_address, 0, sizeof(remote_address));
-  socklen_t remote_addrlen = sizeof(remote_address);
+  std::vector<std::thread> threads;
 
-  std::cout << "Waiting for new connection" << std::endl;
-  int client_socket =
-      accept(sock, (struct sockaddr *)&remote_address, &remote_addrlen);
-  if (client_socket < 0) {
-    perror("Could not accept");
-    return 1;
-  }
-
-  std::string client_ip = inet_ntoa(remote_address.sin_addr);
-  int remote_port = ntohs(remote_address.sin_port);
-
-  std::cout << "Accepted new client @ " << client_ip << ":" << remote_port
-            << std::endl;
-
-  int BUFFLEN = 1024;
-  char buffer[BUFFLEN];
+  std::cout << "Waiting for new connection..." << std::endl;
 
   while (1) {
-    memset(buffer, 0, BUFFLEN);
-    // 5 recv
-    int bytes_received = recv(client_socket, buffer, BUFFLEN - 1, 0);
-    if (bytes_received < 0) {
-      perror("Could not receive");
-      return 1;
-    }
-    if (bytes_received == 0) {
-      std::cout << "Client at " << client_ip << ":" << remote_port
-                << " has disconnected." << std::endl;
-      break;
+    struct sockaddr_in remote_address;
+    memset(&remote_address, 0, sizeof(remote_address));
+    socklen_t remote_addrlen = sizeof(remote_address);
+
+    int client_socket =
+        accept(sock, (struct sockaddr *)&remote_address, &remote_addrlen);
```

Comment:
1) Replace it with the select function with time-out
while (!closeServer) {
        fd_set readfds;
        FD_ZERO(&readfds);
        FD_SET(serverSocket, &readfds);
        
        timeval timeout;
        timeout.tv_sec = 1; // 1 second timeout
        timeout.tv_usec = 0;

        int activity = select(serverSocket + 1, &readfds, NULL, NULL, &timeout);

        if (activity < 0) {
            std::cerr << "Select error\n";
        } else if (activity == 0) {
            // Timeout occurred, check closeServer flag
            continue;
        } else {
            if (FD_ISSET(serverSocket, &readfds)) {
                sockaddr_in clientAddress;
                socklen_t clientSize = sizeof(clientAddress);
                int clientSocket = accept(serverSocket, (sockaddr*)&clientAddress, &clientSize);
                if (clientSocket >= 0) {
                    std::cout << "Client connected!\n";
                    // Start a new thread to handle the client
                    threads.emplace_back(std::thread(handleClient, clientSocket));
                } else {
                    std::cerr << "Failed to accept connection\n";
                }
            }
        }


see https://stackoverflow.com/questions/25888914/linux-c-socket-select-loop
2) In the while loop check if the server got command to quit (Ctrl +C). You can look here
https://stackoverflow.com/questions/1641182/how-can-i-catch-a-ctrl-c-event
And if it set then
exitConnection = true
closeServer = true

This two variable have be atomic, std::atomit<bool>